### ISSUE TITLE
Possible soundness improvements for recent slotmap changes

### ISSUE THREAD

**Issue by detly** [Smithay/calloop#74]
I noticed the changes to use slotmap a while back, and the introduction of a few unsafe code blocks and functions. What mainly caught my eye were a couple of safety notices eg.

> # Safety
> You must ensure that the `*const Token` pointer provided remains alive until your event source is re-registered or unregistered.

(I'll stress that what follows is my analysis, which could be mistaken, so if you're a user take it with a grain of salt until a maintainer of the project provides their feedback.)

The functions in question are marked unsafe, and so there's a "contract" with the caller that they Not Do The Thing you warn against. However, it's not a difficult thing to do, especially when using transient sources (I've certainly done it by mistake myself during development). If it does happen, it's a pretty dire consequence - `epoll()` now has callback data that's effectively a dangling pointer into arbitrary memory.

I think there's a way to change the nature of this, so that instead of a dangling pointer, the result is a leak of a file descriptor and token. Still sub-optimal, but not as bad and technically safe by Rust's contract. Basically it follows the "leak amplification" pattern for other unsafe code in the standard library. An upshot is that it also "pushes back" the unsafe boundary a bit further into the internals of Calloop.

It still uses the same token mechanics, but the initial, invalid token is boxed and converted to a raw pointer as early as possible (see `generic.rs`). The `Poll` type is now responsible for creating a new, valid token at the time of registration, and the source that uses it (probably `Generic`) is responsible for freeing/dropping it. This means that if it is not freed, it leaks, instead of the `Generic` freeing the Box when it's dropped but leaving the pointer in the event loop.

You can take a look at my changes in my [`slotmap-tweaks`](https://github.com/detly/calloop/tree/slotmap-tweaks) branch. They are definitely not finished (as in, they don't even compile yet), because I haven't quite sorted out the type changes required to avoid a leak in the `unregister()` process. If you agree with my analysis, I can open an initial PR and you can review it more fully. If not, I'm happy to discuss it, since I'm certainly not an expert in Rust soundness and safety.

**Comment by vberger** [comment]
Ah, that's interesting, at first glance it seems you raise something pretty relevant, but I don't have the mental bandwidth to review those changes right now. I'll come back to you a little later.

**Comment by detly** [comment]
Yeah no rush, I'm hoping to finish the changes soon, I'll submit a proper PR when I do.

**Comment by detly** [comment]
So I've delved into it some more, and discussed it a bit with the good folk on the Rust Discord (the official one, not the community one). I think that the fundamental problem is that `epoll()` and friends do not "tell" us what the old pointer was that was associated with a file descriptor when we modify or delete an entry. This means that it becomes the responsibility of the caller (in our case, the Rust code) to keep track of it.

This is not, in and of itself, an issue with the API (pretty normal C really). It simply poses a problem in terms of Calloop's architecture combined with Rust's approach to safety. The `Poller` type does not have any knowledge of what event source has what token. It relies entirely on an event source to pass it in. It is therefore possible for an event source to invoke undefined behaviour by messing this up.

The changes I have made so far make this a lot harder to do by accident. However, it's still necessary to mark a couple of the `Poller` functions as `unsafe`, because they can theoretically invoke UB if called from safe code with incorrect arguments.

I think we can go one step further though, and make the `Poller` actually safe. If we keep a mapping of file descriptor to token in `Poller` (I'd suggest a [vec_map](https://docs.rs/vec_map/latest/vec_map/index.html)), then the `Poller` can take responsibility for deallocating the token memory when an event source updates or deletes its registration.

The downside is that (a) it's duplicating information that's already in the kernel ie. `epoll()`'s table of fds to pointers, and (b) it's also kind of duplicating our own token information (or is it? I haven't quite figured that out yet) and (c) it's an extra cost of a few bytes in the struct, and a few more instructions (albeit hardly on a hot path, just on reregistration/unregistration).

Anyway, let me know what you think. I'll try it out regardless and see if it works with the library internals.
